gs_quant/api/fred/fredapi.py

```diff
@@ -0,0 +1,305 @@
+You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+
+//Portions copyright Maverick Lin. Licensed under Apache 2.0 license
+"""
+
+import os
+import sys
+from typing import Iterable, List, Optional, Tuple, Union
+
+import json
+import pandas as pd
+import datetime as dt
+
+import requests
+from requests.utils import requote_uri
+from requests.exceptions import HTTPError
+from urllib.parse import urlencode
+
+"""
+Fred Data API that provides functions to query the Fred dataset. 
+You need to specify a valid API key by passing in the string via api_key. 
+You can sign up for a free API key on the Fred website at http://research.stlouisfed.org/fred2/
+"""
+
+class FredDataApi(object):
+    earliest_realtime_start = '1776-07-04'
+    latest_realtime_end = '9999-12-31'
+    root_url = 'https://api.stlouisfed.org/fred'
+
+    def __init__(self, api_key=None):
+        if api_key is not None:
+            self.api_key = api_key
+        else:
+            import textwrap
+            raise ValueError(textwrap.dedent("""\
+                    Please pass a string with your API key. You can sign up for 
+                    a free api key on the Fred website at 
+                    http://research.stlouisfed.org/fred2/"""))
+
+    def build_query(
+        self,
+        start: Optional[Union[dt.date, dt.datetime]] = None,
+        end: Optional[Union[dt.date, dt.datetime]] = None,
+        as_of: Optional[dt.datetime] = None,
+        since: Optional[dt.datetime] = None,
+        fields: Optional[Iterable[str]] = None,
+        **kwargs
+    ) -> str: 
+        """
+        Builds a FRED URL to query.
+
+        :param start: Requested start date/datetime for data
+        :param end: Requested end date/datetime for data
+        :param as_of: Request data as_of
+        :param since: Request data since
+        :param fields: DataSet fields to include
+        :param kwargs: Extra query arguments
+        :return: a url string of the requested data
+        """
+        
+        if end is not None:
+            assert(isinstance(end, dt.datetime) or isinstance(end, dt.date)), \
+                "End must be of type dt.datetime or dt.date!"
+        if start is not None:
+            assert(isinstance(start, dt.datetime) or isinstance(start, dt.date)), \
+                "Start must be of type dt.datetime or dt.date!"
+        if start is not None and end is not None:
+            if type(start) != type(end):
+                raise ValueError('Start and end types must match!')
+        
+        url = ''
+        date_dict = {
+            "observation_start" : start,
```

Comment:
Please use ' rather than " (unless you need to enclose ') [A few instances of this in this review]